src/tree/gpu_hist/evaluate_splits.cu

```diff
@@ -8,316 +8,269 @@
+  auto scan_input_iter =
+      dh::MakeTransformIterator<ScanElem<GradientSumT>>(
+          bin_iter, ScanValueOp<GradientSumT>{left, right, evaluator});
+
+  dh::device_vector<ReduceElem<GradientSumT>> out_scan(size * 2);
+  auto scan_out_iter = thrust::make_transform_output_iterator(
+      out_scan.begin(), ReduceValueOp<GradientSumT>{left, right, evaluator});
+
+  auto scan_op = ScanOp<GradientSumT>{left, right, evaluator};
+  std::size_t n_temp_bytes = 0;
+  cub::DeviceScan::InclusiveScan(nullptr, n_temp_bytes, scan_input_iter, scan_out_iter,
+                                 scan_op, size * 2);
+  dh::TemporaryArray<int8_t> temp(n_temp_bytes);
+  cub::DeviceScan::InclusiveScan(temp.data().get(), n_temp_bytes, scan_input_iter, scan_out_iter,
+                                 scan_op, size * 2);
+  return out_scan;
+}
 
-  // One block for each feature. Features are sampled, so fidx != blockIdx.x
-  int fidx = inputs.feature_set[is_left ? blockIdx.x
-                                        : blockIdx.x - left.feature_set.size()];
-  if (common::IsCat(inputs.feature_types, fidx)) {
-    EvaluateFeature<BLOCK_THREADS, SumReduceT, BlockScanT, MaxReduceT,
-                    TempStorage, GradientSumT,
-                    OneHotBin<GradientSumT, TempStorage>,
-                    UpdateOneHot<GradientSumT>>(fidx, inputs, evaluator, &best_split,
-                                                &temp_storage);
+template <typename GradientSumT>
+__device__ ScanElem<GradientSumT>
+ScanValueOp<GradientSumT>::MapEvaluateSplitsHistEntryToScanElem(
+    EvaluateSplitsHistEntry entry,
+    EvaluateSplitInputs<GradientSumT> split_input) {
+  ScanElem<GradientSumT> ret;
+  ret.node_idx = entry.node_idx;
+  ret.hist_idx = entry.hist_idx;
+  ret.findex = static_cast<int32_t>(dh::SegmentId(split_input.feature_segments, entry.hist_idx));
+  const bool is_cat = IsCat(split_input.feature_types, ret.findex);
+  if (is_cat || entry.forward) {
+    ret.fvalue = split_input.feature_values[entry.hist_idx];
   } else {
-    EvaluateFeature<BLOCK_THREADS, SumReduceT, BlockScanT, MaxReduceT,
-                    TempStorage, GradientSumT,
-                    NumericBin<GradientSumT, TempStorage, BlockScanT>,
-                    UpdateNumeric<GradientSumT>>(fidx, inputs, evaluator, &best_split,
-                                                 &temp_storage);
+    if (entry.hist_idx > split_input.feature_segments[ret.findex]) {
+      ret.fvalue = split_input.feature_values[entry.hist_idx - 1];
+    } else {
+      ret.fvalue = split_input.min_fvalue[ret.findex];
+    }
   }
-
-  cub::CTA_SYNC();
-
-  if (threadIdx.x == 0) {
-    // Record best loss for each feature
-    out_candidates[blockIdx.x] = best_split;
+  ret.is_cat = is_cat;
+  ret.forward = entry.forward;
+  ret.gpair = split_input.gradient_histogram[entry.hist_idx];
+  ret.parent_sum = split_input.parent_sum;
+  if (((entry.node_idx == 0) &&
```

Comment:
This section of code is hard to understand and could be improved.

Reply:
I will create a small utility function that checks whether `feature_set` contains `ret.findex`.